py4DSTEM/process/phase/parameter_optimize.py

```diff
@@ -0,0 +1,449 @@
+            Dictionaries of arguments to __init__, preprocess, and reconstruct
+            where the OptimizationParameter items have been replaced with the optimal
+            values obtained from the optimizer
+        """
+        init_opt = self._replace_optimized_parameters(
+            self._init_args, self._parameter_list
+        )
+        affine_opt = self._replace_optimized_parameters(
+            self._affine_args, self._parameter_list
+        )
+        affine_transform = partial(AffineTransform, **self._affine_static_args)(
+            **affine_opt
+        )
+        scan_positions = self._get_scan_positions(
+            affine_transform, init_opt["datacube"]
+        )
+        init_opt["initial_scan_positions"] = scan_positions
+
+        prep_opt = self._replace_optimized_parameters(
+            self._preprocess_args, self._parameter_list
+        )
+        reco_opt = self._replace_optimized_parameters(
+            self._reconstruction_args, self._parameter_list
+        )
+
+        return init_opt, prep_opt, reco_opt
+
+    def _replace_optimized_parameters(self, arg_dict, parameters):
+        opt_args = {}
+        for k, v in arg_dict.items():
+            # Check for optimization parameters
+            if isinstance(v, OptimizationParameter):
+                # Find the right parameter in the list
+                # There is probably a better way to do this inverse mapping!
+                for i, p in enumerate(parameters):
+                    if p.name == k:
+                        opt_args[k] = self._skopt_result.x[i]
+            else:
+                opt_args[k] = v
+        return opt_args
+
+    def _split_static_and_optimization_vars(self, argdict):
+        static_args = {}
+        optimization_args = {}
+        for k, v in argdict.items():
+            if isinstance(v, OptimizationParameter):
+                optimization_args[k] = v
+            else:
+                static_args[k] = v
+        return static_args, optimization_args
+
+    def _get_scan_positions(self, affine_transform, dataset):
+        R_pixel_size = dataset.calibration.get_R_pixel_size()
+        x, y = (
+            np.arange(dataset.R_Nx) * R_pixel_size,
+            np.arange(dataset.R_Ny) * R_pixel_size,
+        )
+        x, y = np.meshgrid(x, y, indexing="ij")
+        scan_positions = np.stack((x.ravel(), y.ravel()), axis=1)
+        scan_positions = scan_positions @ affine_transform.asarray()
+        return scan_positions
+
+    def _get_optimization_function(
+        self,
+        cls: type[PhaseReconstruction],
+        parameter_list: list,
+        n_iterations: int,
+        init_static_args: dict,
+        affine_static_args: dict,
+        preprocess_static_args: dict,
+        reconstruct_static_args: dict,
+        init_optimization_params: dict,
+        affine_optimization_params: dict,
+        preprocess_optimization_params: dict,
+        reconstruct_optimization_params: dict,
+    ):
```

Comment:
Adding flags for error divergence and log error.

```suggestion
        check_error_divergence: bool,
        return_log_error:bool,
    ):
```

Reply:
handled by the new custom error metrics